File: app/lib/data/dates.dart
import 'models.dart';

String todayIso() {
  final n = DateTime.now();
  return '${n.year.toString().padLeft(4, '0')}-${n.month.toString().padLeft(2, '0')}-${n.day.toString().padLeft(2, '0')}';
}

DateTime parseIso(String iso) {
  final p = iso.split('-').map(int.parse).toList();
  return DateTime(p[0], p[1], p[2]);
}

String toIso(DateTime d) =>
    '${d.year.toString().padLeft(4, '0')}-${d.month.toString().padLeft(2, '0')}-${d.day.toString().padLeft(2, '0')}';

String addDays(String iso, int days) => toIso(parseIso(iso).add(Duration(days: days)));

String addCycle(String iso, BillingCycle cycle) {
  final d = parseIso(iso);
  return toIso(switch (cycle) {
    BillingCycle.weekly => d.add(const Duration(days: 7)),
    BillingCycle.monthly => DateTime(d.year, d.month + 1, d.day),
    BillingCycle.quarterly => DateTime(d.year, d.month + 3, d.day),
    BillingCycle.yearly => DateTime(d.year + 1, d.month, d.day),
  });
}

int daysUntil(String iso) {
  final target = parseIso(iso);
  final now = parseIso(todayIso());
  return target.difference(now).inDays;
}

String effectiveNextRenewal(Subscription s) {
  if (s.status == SubStatus.cancelled) return s.nextRenewal;
  var date = s.nextRenewal;
  var guard = 0;
  while (daysUntil(date) < 0 && guard < 600) {
    date = addCycle(date, s.cycle);
    guard++;
  }
  return date;
}

String relativeLabel(String iso) {
  final d = daysUntil(iso);
  if (d < 0) return '${d.abs()}d ago';
  if (d == 0) return 'today';
  if (d == 1) return 'tomorrow';
  if (d < 30) return 'in $d days';
  final m = (d / 30).round();
  return m <= 1 ? 'in 1 month' : 'in $m months';
}

File: bubsub/src/lib/dates.ts
import type { BillingCycle, Subscription } from './types'

export function todayISO(): string {
  return toISO(new Date())
}

export function toISO(d: Date): string {
  const y = d.getFullYear()
  const m = String(d.getMonth() + 1).padStart(2, '0')
  const day = String(d.getDate()).padStart(2, '0')
  return `${y}-${m}-${day}`
}

export function parseISO(iso: string): Date {
  const [y, m, d] = iso.split('-').map(Number)
  return new Date(y, m - 1, d)
}

export function addDays(iso: string, days: number): string {
  const d = parseISO(iso)
  d.setDate(d.getDate() + days)
  return toISO(d)
}

export function addCycle(iso: string, cycle: BillingCycle): string {
  const d = parseISO(iso)
  if (cycle === 'weekly') d.setDate(d.getDate() + 7)
  else if (cycle === 'monthly') d.setMonth(d.getMonth() + 1)
  else if (cycle === 'quarterly') d.setMonth(d.getMonth() + 3)
  else d.setFullYear(d.getFullYear() + 1)
  return toISO(d)
}

/** Days from today until the given date (negative if in the past). */
export function daysUntil(iso: string): number {
  const target = parseISO(iso).getTime()
  const now = parseISO(todayISO()).getTime()
  return Math.round((target - now) / 86_400_000)
}

/**
 * Rolls the stored renewal date forward so it is always in the future.
 * Purely computed — we never mutate stored data for display.
 */
export function effectiveNextRenewal(sub: Subscription): string {
  if (sub.status === 'cancelled') return sub.nextRenewal
  let date = sub.nextRenewal
  let guard = 0
  while (daysUntil(date) < 0 && guard < 600) {
    date = addCycle(date, sub.cycle)
    guard++
  }
  return date
}

export function formatDate(iso: string): string {
  return parseISO(iso).toLocaleDateString('en-US', {
    month: 'short',
    day: 'numeric',
    year: 'numeric',
  })
}

export function formatDateShort(iso: string): string {
  return parseISO(iso).toLocaleDateString('en-US', { month: 'short', day: 'numeric' })
}

export function relativeLabel(iso: string): string {
  const days = daysUntil(iso)
  if (days < 0) return `${Math.abs(days)}d ago`
  if (days === 0) return 'Today'
  if (days === 1) return 'Tomorrow'
  if (days < 30) return `in ${days} days`
  const months = Math.round(days / 30)
  return months <= 1 ? 'in 1 month' : `in ${months} months`
}

export function monthName(year: number, month: number): string {
  return new Date(year, month, 1).toLocaleDateString('en-US', { month: 'long', year: 'numeric' })
}
